compiler-core/src/type_/expression.rs

```diff
@@ -691,6 +691,30 @@ impl<'a, 'b> ExprTyper<'a, 'b> {
                 unify(left.type_(), right.type_())
                     .map_err(|e| convert_unify_error(e, right.location()))?;
 
+                match (&left, &right) {
```

Comment:
Let's extract this into a new function, to make it easier to read.

There's also a very large amount of nesting in this function- let's rewrite it in a flat style using early returns.

Reply:
Done!